lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h

```diff
@@ -9,31 +9,68 @@
 namespace mlir {
 namespace heir {
 
+/// This class represents the multiplicative depth in an FHE program.
+///
+/// The multiplicative depth of a circuit refers to the maximum number of
+/// multiplication gates on any path from a circuit input to an output. In a
+/// more general FHE program, the multiplicative depth is the maximum number of
+/// ciphertext-ciphertext or plaintext-ciphertext multiplication operations on
+/// any use-def chain.
+///
+/// In FHE, multiplicative depth bounds the number of operations that can be
+/// sequentially performed on a ciphertext before the noise introduced during
+/// those operations renders the ciphertext undecryptable.
+///
+/// For example, the multiplicative depth of the following pseudo-code is 2:
+///
+///     %0 = arith.constant 0 : multiplicative depth = 0
+///     %1 = FHE.mul %0 %0 : multiplicative depth = 1
+///     %2 = FHE.add %0 %1 : multiplicative depth = 1
+///     %3 = FHE.mul %2 %0 : multiplicative depth = 2
+///
+/// In more detail, the multiplicative depth for an arbitrary operation `op`
+/// can be defined inductively as follows.
+///
+///     MulDepth(z) = case
+///       mul_op(x, y):
+///         max(MulDepth(x), MulDepth(y)) + 1
+///       any_op(operands):
+///         max(map(MulDepth, operands))
+
 class MulDepth {
  public:
   MulDepth() : value(std::nullopt) {}
   explicit MulDepth(int64_t value) : value(value) {}
   ~MulDepth() = default;
 
+  /// Whether the multiplicative depth is initialized. It can be uninitialized
+  /// when the state hasn't been set during the analysis.
   bool isInitialized() const { return value.has_value(); }
 
+  /// Get a known multiplicative depth.
   const int64_t &getValue() const {
     assert(isInitialized());
     return *value;
   }
 
   bool operator==(const MulDepth &rhs) const { return value == rhs.value; }
 
+  /// Join two multiplicative depth.
   static MulDepth join(const MulDepth &lhs, const MulDepth &rhs) {
     if (!lhs.isInitialized()) return rhs;
     if (!rhs.isInitialized()) return lhs;
+    // If both are initialized, the larger value should be chosen for a sound
+    // analysis of the multiplicative depth. If both are initialized, the larger
+    // value between the two should be selected to soundly analyze the
+    // multiplicative depth.
```

Comment:
Maybe this was a copy-paste thing?

```suggestion
    // If both are initialized, the larger value should be chosen for a sound
    // analysis of the multiplicative depth.
```

Reply:
Oops!

Reply:
@asraa Could you possibly squash merge this commit? Otherwise, I can do it after 12 hours :)

Reply:
done!